src/diagram/Timeline.ts

```diff
@@ -0,0 +1,92 @@
+import { Cancellable } from "../command/Commands";
+import { Diagram, Memento } from "./Diagram";
+
+export class Snapshot<T extends Cancellable> {
+  constructor(readonly origin: Memento, readonly modifier: T) {}
+}
+
+export class Timeline<T extends Cancellable> {
+  private diagram!: Diagram | null;
+  private latest!: Memento;
```

Comment:
why use an exclamation mark here?

Reply:
Since Typescript cannot identify the resetHistory function can initialize the variable latest in the constructor. The ! mark is needed for here